### PR TITLE
fix #12150 C++ free functions are now supported using new `importcpp:"!$1"` syntax

### LINKED ISSUE #12150: `importcpp` does not work with C++ backend
Doesn't look like importcpp works at all, at least for free functions.

### Example
```nim
# t0725.nim
proc fun1() {.importcpp.}
fun1()
import ./t0725b

# t0725b.nim
{.emit:"""void fun1(){}""".}
# also fails with this:
# proc fun1() {.exportcpp.} = discard
```
```
nim cpp -r t0725.nim
```
### Current Output
```
t07725.nim.cpp:118:2: error: use of undeclared identifier 'fun1'
```

### Expected Output
should work

### Additional Information
* Your Nim version (output of `nim -v`).
latest devel e0f3e8432af6a64d01be089a2247565e1f754b3a
on OSX, but shouldn't matter

* Was it working in the previous Nim releases?
no

### PR DESCRIPTION
**Pull request by timotheecour**
fix https://github.com/nim-lang/Nim/issues/12150

## example
```nim
proc freeFn(a: cint) {.importcpp: "!$1".}
# maps to `void freeFn(int)`
```

### CODE REVIEW

**Review by Clyybber — COMMENTED** [review]
This should be handled in `ccgcalls.genPatternCall` instead.

File: compiler/ast.nim [hunk 1]
```diff
@@ -295,6 +295,9 @@ type
     sfUsedInFinallyOrExcept  # symbol is used inside an 'except' or 'finally'
     sfSingleUsedTemp  # For temporaries that we know will only be used once
     sfNoalias         # 'noalias' annotation, means C's 'restrict'
+    sfCompileToCpp
```
Comment (Araq) [inline compiler/ast.nim:301]: I don't understand this. You might as well remove the ``sfCompileToCpp`` alias.

Comment (timotheecour) [reply compiler/ast.nim:301]: before PR `sfCompileToCpp` and `sfInfixCall` were aliases, but after PR they're not aliases and I need the distinction to support C++ free functions, which fixes #12150; I've went through each use of sfCompileToCpp vs sfInfixCall to make sure the correct one is used in each case.

to further clarify:
* `sfCompileToCpp` is still set for all C++ importcpp procs
* `sfInfixCall` now isn't set for a C++ free function (which is indeed not an infix call)

**Comment by timotheecour** [comment]
> This should be handled in ccgcalls.genPatternCall instead.

this wouldn't help with #12150, try it; one problem was that `sfInfixCall` inhibited emitting a declaration in client code that calls `importcpp` on a free function

**Comment by Varriount** [comment]
Why doesn't `importc` work for this?

**Comment by timotheecour** [comment]
> Why doesn't importc work for this?

because it would use c mangling. If you have a C++ free function, (eg: `void fun(int a);`), this would mangle differently.

**Comment by Araq** [comment]
> this wouldn't help with #12150, try it; one problem was that sfInfixCall inhibited emitting a declaration in client code that calls importcpp on a free function

Please explain this in more detail.

**Comment by timotheecour** [comment]
@araq this line should explain everything:
```
if sfInfixCall notin prc.flags: genProcPrototype(m, prc)
```
in the example of https://github.com/nim-lang/Nim/issues/12150, `fun1` is an importcpp proc therefore before PR has `sfInfixCall in prc.flags`, so `genProcPrototype` isn't called, and `fun1` isn't declared in t0725.nim, hence giving `error: use of undeclared identifier 'fun1'`

after PR, I distinguish C++ free functions (which need a declaration) vs methods (`sfInfixCall`), which don't.

**Comment by Araq** [comment]
Ok, but I meant to ask:

> This should be handled in ccgcalls.genPatternCall instead.

Why can't it be done in genPatternCall? Seems the most obvious place to put it.

**Comment by timotheecour** [comment]
I don't see how that would work, at very least I don't see how it would be simpler than the way I wrote it, if you have a concrete idea in mind please provide more details / edit this PR.

I need to tell whether an importcpp symbol is sfInfixCall or not early on (inside `processImportCpp`) since it affects semantics/codegen in a few places, so moving the following lines to `genPatternCall` adds complexity
```nim
let isFreeFunction = extname.startsWith "!"
  if isFreeFunction:
    extname = extname[1..^1]
  else:
    incl(s.flags, sfInfixCall)
```
and interface of `genPatternCall` doesn't fit well here

**Comment by timotheecour** [comment]
Hi @Araq anything else needed here? I rebased to fix conflicts

**Comment by timotheecour** [comment]
rebased again to fix bitrot conflicts... I'd really like to have this merged

**Comment by timotheecour** [comment]
ping @xflywind

File: compiler/pragmas.nim [hunk 2]
```diff
@@ -175,9 +175,15 @@ proc processImportCompilerProc(c: PContext; s: PSym, extname: string, info: TLin
   incl(s.loc.flags, lfImportCompilerProc)
 
 proc processImportCpp(c: PContext; s: PSym, extname: string, info: TLineInfo) =
+  var extname = extname
+  let isFreeFunction = extname.startsWith "!"
```
Comment (ringabout) [inline compiler/pragmas.nim:182]: add some comments here, clearifying the pattern of free function

**Review by ringabout — APPROVED** [review]
looks good

File: doc/manual.rst [hunk 3]
```diff
@@ -6947,18 +6947,22 @@ language for maximum flexibility:
 - A dot following the hash ``#.`` indicates that the call should use C++'s dot
   or arrow notation.
 - An at symbol ``@`` is replaced by the remaining arguments, separated by commas.
+- An exclamation symbol ``!`` indicates a free function.
```
Comment (ringabout) [inline doc/manual.rst]: please explain more, it is not easy to understand "free function".

https://stackoverflow.com/questions/4861914/what-is-the-meaning-of-the-term-free-function-in-c

File: compiler/pragmas.nim (hunk 2 above)
Comment (timotheecour) [reply compiler/pragmas.nim:182]: done

File: doc/manual.rst (hunk 3 above)
Comment (timotheecour) [reply doc/manual.rst]: added meaning as non-member (and there are examples right below)

**Review by timotheecour — COMMENTED** [review]
PTAL

**Comment by timotheecour** [comment]
@Araq any other comments or can we merge this?

**Comment by Araq** [comment]
Sorry for being late but if we really need this, we should have a dedicated `importCppProc` pragma instead of more magic going on in the import pattern.

**Comment by timotheecour** [comment]
maybe we need `{.pattern: .}` instead of this PR, which would be more general (work in c,cpp,js backends), refs https://github.com/nim-lang/RFCs/issues/315#issuecomment-873171898

eg:
```nim
proc freeFn(a: cint) {.pattern: "$call".} # whole call; honors backend mangling (c/c++/js)
proc freeFn2(a, b: cint) {.pattern: "std::bar(@)".} # calls std::bar(a, b)
proc `+`*(a, b: int128): int128 {.pattern: "$1 + $2".} # calls a + b
proc `+!`*(a, b: int128): int128 {.pattern: "$2 + $1".}  # calls b + a
proc freeFn3(a, b: cint) {.pattern: "$0(1, @)".} # calls freeFn3(1, a, b)
```

**Comment by Araq** [comment]
Either `importCppProc` or `pattern` would be better. Closing for now.

**Pull request closed without merging**